plugin/firewall/policy/expression.go

```diff
@@ -0,0 +1,120 @@
+package policy
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/coredns/coredns/plugin/pkg/rqdata"
+
+	"github.com/coredns/coredns/request"
+
+	"github.com/coredns/coredns/plugin/metadata"
+
+	expr "github.com/Knetic/govaluate"
+)
+
+type ruleExpr struct {
+	kind        int
+	kindIfError int
+	expression  *expr.EvaluableExpression
+}
+
+// ExprEngine implement interface Engine for Firewall plugin
+// it evaluate the rues using an the lib Knetic/govaluate
+type ExprEngine struct {
+	kindIfErrorEvaluation int
+	dataFromReq           *rqdata.Mapping
+}
+
+type dataAsParam struct {
+	ctx         context.Context
```

Comment:
why does this need to embed a context?

Reply:
dataAsParam is the Evaluator for variable that are embeded in the expression.
Those variables can be either a name of a field of the request (resolved by the RqData), either a Label of Metadata. We need the context to resolve the Metadata labels.

Reply:
It's very odd to put the ctx in the struct; I don't we do it currently. Should just give it the functions you're calling